components/fxa-client/ios/FxAClient/FxAccountDeviceConstellation.swift

```diff
@@ -0,0 +1,192 @@
+    // Refresh the list of remote devices.
+    // Dispatchs a notification to NSNotificationCenter, using the name `.constellationStateUpdate`.
+    // The updated `ConstellationState` can be found in `userInfo` under the key `newState`.
+    func refreshState()
+
+    func setLocalDeviceName(name: String)
+
+    // Poll for events we might have missed (e.g. no push notification)
+    func pollForEvents(completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void)
+    // Send an event to another device such as Send Tab.
+    func sendEventToDevice(targetDeviceId: String, e: DeviceEventOutgoing)
+
+    // Register our push subscription with the FxA server.
+    func setDevicePushSubscription(sub: DevicePushSubscription)
+    // Used by Push when receiving a message.
+    func processRawIncomingDeviceEvent(
+        pushPayload: String,
+        completionHandler: @escaping (Result<[DeviceEvent], Error>) -> Void
+    )
+}
+
+public extension Notification.Name {
+    static let constellationStateUpdate = Notification.Name("constellationStateUpdate")
+}
+
+public struct ConstellationState {
+    public let localDevice: Device?
+    public let remoteDevices: [Device]
+}
+
+public class DeviceConstellation: DeviceConstellationProtocol {
+    var constellationState: ConstellationState?
+    let account: FirefoxAccount
+
+    init(account: FirefoxAccount) {
+        self.account = account
+    }
+
+    public func state() -> ConstellationState? {
+        return constellationState
+    }
+
+    public func refreshState() {
+        DispatchQueue.global().async {
+            FxALog.info("Refreshing device list...")
+            do {
+                let devices = try self.handleAccountAuthErrors { acct in
+                    try acct.fetchDevicesSync()
+                }
+                let localDevice = devices.first { $0.isCurrentDevice }
+                if localDevice?.subscriptionExpired ?? false {
+                    FxALog.debug("Current device needs push endpoint registration.")
+                }
+                let remoteDevices = devices.filter { !$0.isCurrentDevice }
+
+                let newState = ConstellationState(localDevice: localDevice, remoteDevices: remoteDevices)
+                self.constellationState = newState
+
+                FxALog.debug("Refreshed device list; saw \(devices.count) device(s).")
+
+                DispatchQueue.main.async {
+                    NotificationCenter.default.post(
+                        name: .constellationStateUpdate,
+                        object: nil,
+                        userInfo: ["newState": newState]
+                    )
+                }
+            } catch {
+                FxALog.error("Failure fetching the device list: \(error).")
+                return
+            }
+        }
+    }
+
+    func initDevice(name: String, type: DeviceType, capabilities: [DeviceCapability]) {
+        // This is already wrapped in a `fxaQueue.async`, no need to re-wrap.
```

Comment:
These two comments appear to be the only two mentions of `fxaQueue.async` in the PR, so I'm not entirely sure what you mean here. Is it because they've invoked by the accountmanager in its own already-async method?

Reply:
(In which case, I wonder if there is some way to ensure they won't accidentally be called by other things that do not provide the same property).